### PR TITLE
Cudnn convolution reconstruction 

### PR DESCRIPTION
**Pull request by wangchaochaohu**
cudnn convolution reconstruction using conv_cudnn_helper.h

1: add half_float for the conv grad
2: fix some bug in the conv_cudnn_helper(the usage of cudnnGetConvolutionBackwardDataWorkspaceSize)
3: reconstruction the conv_cudnn_helper(both v7 algorithm and the workspace size limit)
This PR will solve goognet training nan problem and illegal problem, It also sovle the problem of mask-fpn model training problem.

### CODE REVIEW

File: paddle/fluid/operators/conv_cudnn_op.cu.cc [hunk 1]
```diff
@@ -361,263 +205,83 @@ class CUDNNConvGradOpKernel : public framework::OpKernel<T> {
...
-      T* filter_grad_data = filter_grad->mutable_data<T>(ctx.GetPlace());
       // Because beta is zero, it is unnecessary to reset filter_grad.
       for (int i = 0; i < groups; i++) {
-        CUDNN_ENFORCE(platform::dynload::cudnnConvolutionBackwardFilter(
-            handle, &alpha, cudnn_input_desc, input_data + i * group_offset_in,
-            cudnn_output_grad_desc, output_grad_data + i * group_offset_out,
-            cudnn_conv_desc, filter_algo, cudnn_workspace_ptr,
-            workspace_size_in_bytes, &beta, cudnn_filter_desc,
-            filter_grad_data + i * group_offset_filter));
+        workspace_handle.RunFunc(
```
Comment (qingqing01) [inline paddle/fluid/operators/conv_cudnn_op.cu.cc:274]: Here use `workspace_handle.RunFunc` not  ctx.AllocateTmpTensor to get cuDNN worksapce,  please ask @chengduoZH @sneaxiy to review.

**Review by qingqing01 — APPROVED** [review]

File: paddle/fluid/operators/conv_cudnn_helper.h [hunk 2]
```diff
@@ -128,15 +169,72 @@ struct SearchAlgorithm<cudnnConvolutionBwdDataAlgoPerf_t> {
...
+      std::unique_ptr<perf_t[]> perf_results(
+          new perf_t[kNUM_CUDNN_BWD_DATA_ALGS]);
+      CUDNN_ENFORCE(
+          platform::dynload::cudnnGetConvolutionBackwardDataAlgorithm_v7(
+              args.handle, args.wdesc.desc(), args.odesc.desc(),
+              args.cdesc.desc(), args.idesc.desc(), kNUM_CUDNN_BWD_DATA_ALGS,
+              &perf_count, perf_results.get()));
+      algo = (perf_results.get())[best_algo_idx].algo;
+
+#if CUDNN_VERSION < 7500
```
Comment (chengduoZH) [inline paddle/fluid/operators/conv_cudnn_helper.h:203]: Why do you add `if CUDNN_VERSION < 7500` here?

**Comment by chengduoZH** [comment]
> 2: fix some bug in the conv_cudnn_helper

Can you describe this bug?

**Review by sneaxiy — APPROVED** [review]
LGTM. We should monitor kpis of benchmark later after this PR.

File: paddle/fluid/operators/conv_cudnn_helper.h (hunk 2 above)
Comment (wangchaochaohu) [reply paddle/fluid/operators/conv_cudnn_helper.h:203]: nv fix the bug above cudnn 7.5, wo when cudnn_verison > 7500, we need not to check the blacklist

**Comment by wangchaochaohu** [comment]
> > 2: fix some bug in the conv_cudnn_helper
> 
> Can you describe this bug?

Done

**Pull request merged**